### DISCUSSION TITLE
RGB PDF - Replace color and convert to CMYK PDF?

### DISCUSSION THREAD

**Discussion by christianslater** [ImageMagick/ImageMagick#2766] [Digital Image Processing]
Hello community, 
hopefully this is not a dumb question but I am totally new to imagemagick.

I have a online-design-tool that saves the design as RGB PDF file. I guess TCPDF is used for that. Unfortunatelly I need this file as CMYK PDF for printing. Is this possible with imagemagick?

The Second question is more specific. In some designs we want to print a specific color. For that the printing company needs the PDF to be prepared in a special way. Is it possible with imagemagick to find all elements with a specific HEX color value and replace this color with a spot color with a custom name and specific CMYK values?

Thanks in advance for Your reply!

**Comment by snibgo** [comment]
IM is a raster image processor. When it writes PDF, it embeds a raster image inside the PDF.

If your RGB PDF files are also raster, then you can extract the raster image with pdfimages and then IM can do any processing and write the CMYK raster image to a PDF.

IM can replace any colour with any other colour, but not a "spot color with a custom name".

But if your RGB PDF files are vector, IM will rasterize them first. Then IM can do any processing, and write the CMYK raster image to a PDF. It won't write a CMYK vector image.

If you are uncertain whether your RGB PDF files are vector or raster, you can post a link to one here.

**Reply by christianslater** [reply]
Hello snibgo,
I really appreciate your answer!

Generally The elements in the RGB PDF where I need the color replacement to be done are Texts or Vectors. I attached a sample PDF. 

So when I get You right, I can replace this color with IM and generate a CMYK PDF with IM but **can not** create a spot color?

Is there any tool that can do this?

Is there any place where I can find a developer that can realize this for me?
[3731_100.pdf](https://github.com/ImageMagick/ImageMagick/files/5446373/3731_100.pdf)

**Reply by t246246** [reply]
If pdf is enough simple, and only need to change text and some lines, one approach is to use qpdf.
First, conver pdf to qdf which is almost consisted from ascii text and one can edit it using text editor.

% qpdf -qdf input.pdf output.qdf

After finishing edit, you can back it to pdf.

% fix-qdf < output_edited.qdf > input_edited.pdf

I see your pdf, it contains text and some lines.

In text part it looks like

0.000000 0.000000 0.000000 rg
0.800000 0.643137 0.113725 rg

where rgb color is set.  So replace all occurrence of rg operator to corresponding cmyk color setting operator k:

0 0 0 1 k
0 0.19607875 0.85784375 0.2 k

I see another RGB element is used in default colorspace.

 /Group <<
    /CS /DeviceRGB
    /S /Transparency
    /Type /Group
  >>

And it is used like this:

/CS1 CS 1.000000 SCN

This case, just replace the default seems to be enough.

So, replace it to

 /Group <<
    /CS /DeviceCMYK
    /S /Transparency
    /Type /Group
  >>

I hope this edit is enough for your file.  Well, using this approach, you need to some knowledge to PDF itself, though.  But If all your PDF is made from same application, this would be your help.

**Reply by christianslater** [reply]
thanks for this approach! Generally most of the time it is just text or simple graphics. So each time the spefified hex value is appearing it must be converted.

Does qpdf support solid colors with custom names?

**Reply by t246246** [reply]
One of the job of qpdf is to rewirte to the readable one.  You can rewrite it to any pdf  as long as you know the correct grammar of pdf.
Here is a sample pdf using various colors https://www.pdflib.com/fileadmin/cookbooks/PDFlib/results/starter_color.pdf
Convert it to visible one using qpdf, you can see how to write spotname in pdf.

However it is not so easy job in general. See https://stackoverflow.com/questions/2022338/how-can-i-change-a-specific-rgb-color-into-a-specific-cmyk-color-in-a-pdf-file for example.  There may be a commercial product to do it but I don't know how good or bad the conversion is.

**Comment by snibgo** [comment]
> So when I get You right, I can replace this color with IM and generate a CMYK PDF with IM but can not create a spot color?

IM knows nothing about spot colours. (A spot colour is usually a special ink, such as a metallic colour, that cannot be reproduced in the CMYK colour model.)

IM can generate CMYK PDF files. BUT these will not contain vector or text elements. Instead, they will contain raster images. (A raster image has pixels, not vectors.)

You might look at Open Office https://extensions.openoffice.org/en/node/17351 or Libre Office.

Adobe also has a tool for editing PDF files. It isn't free, but a free trial is available.

I don't have much experience of editing PDF files.

**Reply by christianslater** [reply]
Hi,
generally the printing company need the color as solid color with the value 0 100 0 0 (CMYK) and the name “goldfoil”. Visually it is just a magenta color but as it is not a processcolor and has this special name, they know to replace it with goldfoil.

It is no problem when it is rastered.

Does IM support solid color with custom name?

I am designer and know how to do it with Illustrator and co... but I need this process to be automated running on a server

**Comment by snibgo** [comment]
> Does IM support solid color with custom name?

Not directly. This might be possible in a raster image by specifying some metadata. Do you have a sample raster file (.PNG, .TIFF etc) with a spot colour?

**Comment by christianslater** [comment]
Hi snibgo, I attached a PDN (Tiff was too large) and a jpg. I think in rastered images there is no custom name to the color to be attached. Just the values. In my Case 0 100 0 0 (CMYK) In the PDF the color is stored with it's custom name. See screenshot.
![screenshot_customcolor](https://user-images.githubusercontent.com/1314513/97772577-c9131b80-1b48-11eb-8617-1ef508bb69cb.jpg)
[Unbenannt-1.pdf](https://github.com/ImageMagick/ImageMagick/files/5468606/Unbenannt-1.pdf)
![Unbenannt-1](https://user-images.githubusercontent.com/1314513/97772621-2018f080-1b49-11eb-8a79-fe10ad289097.jpg)

**Reply by t246246** [reply]
As far as I understand spot color is a concept of PostScript and PDF.  Some format may have the same concept, but I don't think raster image does not have it.

**Comment by snibgo** [comment]
In that PDF and JPG, neither exiftool nor IM "-verbose info:" shows any metadata about spot colours.

**Reply by christianslater** [reply]
Mhm. In acrobat pro the pdf shows the spot color. I’ve seen that TCPDF can produce spot color but I am not sure if its possible to replace an existing color

**Reply by t246246** [reply]
Because spot color is an object of PDF, not in metadata.
In your Unbenannt-1.pdf, it is here:
```
8 0 obj
[/Separation/goldfolie/DeviceCMYK<</C0[0.0 0.0 0.0 0.0]/C1[0.0 1.0 0.0 0.0]/Domain[0 1]/FunctionType 2/N 1.0/Range[0.0 1.0 0.0 1.0 0.0 1.0 0.0 1.0]>>]
endobj
```
